### PR TITLE
planner/core: fix a bug during add cast for decimal join key 

### PR DESCRIPTION
**Pull request by hanfei1991**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tics/issues/1704

Problem Summary:

The cause of this bug is that after [pr convert decimal type join key](https://github.com/pingcap/tidb/pull/23191) is merged, we will change the decimal join key with a well-casted one. But when we construct the physical, the join keys is shallow copied. Changing join keys will change the keys on other hash join candidates and results in this bug. 

### What is changed and how it works?

What's Changed:
deep copy the join equal expressions when planner replace the join key with the cast function.

### Check List 

Tests 

- Unit test

### Release note 

- planner/core: fix a bug during add cast for decimal join key

### CODE REVIEW

**Review by qw4990 — APPROVED** [review]

File: executor/tiflash_test.go [hunk 1]
```diff
@@ -210,6 +210,8 @@ func (s *tiflashTestSuite) TestMppExecution(c *C) {
 	tk.MustQuery("select t1.c1 from t t1 join t t2 on t1.c1 = t2.c1 order by t1.c1").Check(testkit.Rows("1.00000", "1.00001", "1.00010"))
 	tk.MustQuery("select t1.c1 from t t1 join t t2 on t1.c1 = t2.c3 order by t1.c1").Check(testkit.Rows("1.00000", "1.00000", "1.00010"))
 	tk.MustQuery("select t1.c4 from t t1 join t t2 on t1.c4 = t2.c3 order by t1.c4").Check(testkit.Rows("1.0000", "1.0000", "1.0001"))
+	// let this query choose hash join
+	tk.MustQuery("select /*+ nth_plan(2) */ t1.c1 from t t1 join t t2 on t1.c1 = t2.c3 order by t1.c1").Check(testkit.Rows("1.00000", "1.00000", "1.00010"))
```
Comment (fzhedu) [inline executor/tiflash_test.go:214]: /*+ nth_plan(2) */ is stable to guarantee choose non-mpp plan?

**Review by fzhedu — APPROVED** [review]
/LGTM

File: executor/tiflash_test.go (hunk 1 above)
Comment (hanfei1991) [reply executor/tiflash_test.go:214]: I think so , there is only one type of mpp join for now.

**Pull request merged**